In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
#import data
content_df = pd.read_csv('/content/Content.csv',index_col=0)
content_df.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   
2  230c4e4d-70c3-461d-b42c-ec09396efb3f  a5c65404-5894-4b87-82f2-d787cbee86b4   
3  356fff80-da4d-4785-9f43-bc1261031dc6  9fb4ce88-fac1-406c-8544-1a899cee7aaf   
4  01ab84dd-6364-4236-abbb-3f237db77180  e206e31b-5f85-4964-b6ea-d7ee5324def1   

    Type        Category                                                URL  
0  photo        Studying  https://socialbuzz.cdn.com/content/storage/975...  
1  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/9f7...  
2  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/230...  
3  photo      technology  https://socialbuzz.cdn.com/content/storage/356...  
4  video            food  https://socialbuzz.cdn.com/content/storage/01a...

In [ ]:
content_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  1000 non-null   object
 1   User ID     1000 non-null   object
 2   Type        1000 non-null   object
 3   Category    1000 non-null   object
 4   URL         801 non-null    object
dtypes: object(5)
memory usage: 46.9+ KB


In [ ]:
#Summary of the data
content_df.describe()

Content ID  \
count                                   1000   
unique                                  1000   
top     97522e57-d9ab-4bd6-97bf-c24d952602d2   
freq                                       1   

                                     User ID   Type    Category  \
count                                   1000   1000        1000   
unique                                   446      4          41   
top     72d2587e-8fae-4626-a73d-352e6465ba0f  photo  technology   
freq                                       8    261          71   

                                                      URL  
count                                                 801  
unique                                                801  
top     https://socialbuzz.cdn.com/content/storage/975...  
freq                                                    1

* We won't need the User ID or URL columns for our analysis, so let's drop those. Then let's check the values in the category column to make sure they make sense given our data.

In [ ]:
content_df.drop(labels=['User ID','URL'],axis=1,inplace=True)

In [ ]:
content_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  1000 non-null   object
 1   Type        1000 non-null   object
 2   Category    1000 non-null   object
dtypes: object(3)
memory usage: 31.2+ KB


In [ ]:
#Check values in category columns
content_df['Category'].value_counts()

technology           71
animals              67
travel               67
culture              63
science              63
fitness              61
food                 61
healthy eating       61
cooking              60
soccer               58
tennis               58
education            57
dogs                 56
studying             55
veganism             48
public speaking      48
Fitness               5
Animals               4
Science               4
"soccer"              3
"culture"             3
Soccer                3
"dogs"                2
Education             2
Studying              2
Travel                2
Food                  2
"veganism"            1
"public speaking"     1
Public Speaking       1
"technology"          1
"cooking"             1
Healthy Eating        1
"studying"            1
"food"                1
Culture               1
"tennis"              1
Technology            1
"animals"             1
Veganism              1
"science"             1
Name: Category, 

* We'll need to clean up these categories by removing punctuation and capitalization.

In [ ]:
#strip quote mark and lowercase all values
content_df['Category'] = content_df['Category'].apply(lambda x:x.strip('"').lower())

In [ ]:
#checking if all categories are in lower case
content_df['Category'].value_counts()

technology         73
animals            72
travel             69
science            68
culture            67
fitness            66
food               64
soccer             64
healthy eating     62
cooking            61
tennis             59
education          59
studying           58
dogs               58
public speaking    50
veganism           50
Name: Category, dtype: int64

In [ ]:
#rename content to content_type
content_df.rename(columns={'Type':'Content_Type'},inplace=True)
content_df['Content_Type'].value_counts()

photo    261
video    259
GIF      244
audio    236
Name: Content_Type, dtype: int64

In [ ]:
#import 2nd dataset
reactions_df = pd.read_csv('/content/Reactions.csv',index_col=0)
reactions_df.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2                                   NaN   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   
3  97522e57-d9ab-4bd6-97bf-c24d952602d2  163daa38-8b77-48c9-9af6-37a6c1447ac2   
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  34e8add9-0206-47fd-a501-037b994650a2   

      Type             Datetime  
0      NaN  2021-04-22 15:17:15  
1  disgust  2020-11-07 09:43:50  
2  dislike  2021-06-17 12:22:51  
3   scared  2021-04-18 05:13:58  
4  disgust  2021-01-06 19:13:01

In [ ]:
reactions_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25553 entries, 0 to 25552
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  25553 non-null  object
 1   User ID     22534 non-null  object
 2   Type        24573 non-null  object
 3   Datetime    25553 non-null  object
dtypes: object(4)
memory usage: 998.2+ KB


In [ ]:
reactions_df.describe()

Content ID  \
count                                  25553   
unique                                   980   
top     4b2d0fff-3b4f-43ca-a7df-c430479cb9ba   
freq                                      49   

                                     User ID   Type             Datetime  
count                                  22534  24573                25553  
unique                                   500     16                25542  
top     c76c3393-88e2-47b0-ac37-dc4f2053f5a5  heart  2020-10-29 20:51:08  
freq                                      65   1622                    2

* Again, we can drop the User ID column. We'll also need to drop rows with missing values, and change the "Datetime" column to a datetime data type. Finally, we'll need to change the "Type" column name like we did for the Content table.



In [ ]:
# drop user ID column
reactions_df.drop(labels = ['User ID'],axis=1,inplace=True)

In [ ]:
# convert dates to datetime
reactions_df['Datetime'] = pd.to_datetime(reactions_df['Datetime'])


In [ ]:
#drop rows with missing values
reactions_df.dropna(inplace=True)

In [ ]:
#rename column
reactions_df.rename(columns={"Type":'Reaction_Type'},inplace=True)

In [ ]:
reactions_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24573 entries, 1 to 25552
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Content ID     24573 non-null  object        
 1   Reaction_Type  24573 non-null  object        
 2   Datetime       24573 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(2)
memory usage: 767.9+ KB


In [ ]:
#import 3rd Dataset
reactiontypes_df = pd.read_csv('/content/ReactionTypes.csv',index_col=0)
reactiontypes_df.head()

Type Sentiment  Score
0       heart  positive     60
1        want  positive     70
2     disgust  negative      0
3        hate  negative      5
4  interested  positive     30

In [ ]:
reactiontypes_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16 entries, 0 to 15
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Type       16 non-null     object
 1   Sentiment  16 non-null     object
 2   Score      16 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 512.0+ bytes


In [ ]:
reactiontypes_df.describe()

Score
count  16.000000
mean   39.625000
std    26.901983
min     0.000000
25%    14.250000
50%    40.000000
75%    66.250000
max    75.000000

In [ ]:
#rename type column
reactiontypes_df.rename(columns={'Type':'Reaction_Type'},inplace=True)
reactiontypes_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16 entries, 0 to 15
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Reaction_Type  16 non-null     object
 1   Sentiment      16 non-null     object
 2   Score          16 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 512.0+ bytes


In [ ]:
reactiontypes_df.value_counts()

Reaction_Type  Sentiment  Score
adore          positive   72       1
cherish        positive   70       1
disgust        negative   0        1
dislike        negative   10       1
hate           negative   5        1
heart          positive   60       1
indifferent    neutral    20       1
interested     positive   30       1
intrigued      positive   45       1
like           positive   50       1
love           positive   65       1
peeking        neutral    35       1
scared         negative   15       1
super love     positive   75       1
want           positive   70       1
worried        negative   12       1
dtype: int64

* It looks like our data is clean, based on the client's specifications (remove rows with missing values, drop unnecessary colunmns, and ensure data types match the column values).

# Creating a joined table for analysis (Data Modelling)

In [ ]:
#merge tables on content ID columns
top_performers = reactions_df.merge(content_df,on='Content ID')
top_performers

Content ID Reaction_Type            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike 2021-06-17 12:22:51   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2        scared 2021-04-18 05:13:58   
3      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-01-06 19:13:01   
4      97522e57-d9ab-4bd6-97bf-c24d952602d2    interested 2020-08-23 12:25:58   
...                                     ...           ...                 ...   
24568  75d6b589-7fae-4a6d-b0d0-752845150e56       dislike 2020-06-27 09:46:48   
24569  75d6b589-7fae-4a6d-b0d0-752845150e56     intrigued 2021-02-16 17:17:02   
24570  75d6b589-7fae-4a6d-b0d0-752845150e56    interested 2020-09-12 03:54:58   
24571  75d6b589-7fae-4a6d-b0d0-752845150e56       worried 2020-11-04 20:08:31   
24572  75d6b589-7fae-4a6d-b0d0-752845150e56       cherish 2021-01-04 04:55:11   

      Content_Type    Category  
0            photo    studying  
1            photo    studying  
2            photo    studying  
3            photo    studying  
4            photo    studying  
...            ...         ...  
24568        audio  technology  
24569        audio  technology  
24570        audio  technology  
24571        audio  technology  
24572        audio  technology  

[24573 rows x 5 columns]

In [ ]:
#merge tables on reaction type column
top_performers = top_performers.merge(reactiontypes_df,on='Reaction_Type')
top_performers

Content ID Reaction_Type            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-01-06 19:13:01   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-04-09 02:46:20   
3      9f737e0a-3cdd-4d29-9d24-753f4e3be810       disgust 2021-03-28 21:15:26   
4      230c4e4d-70c3-461d-b42c-ec09396efb3f       disgust 2020-08-04 05:40:33   
...                                     ...           ...                 ...   
24568  435007a5-6261-4d8b-b0a4-55fdc189754b         adore 2020-10-04 22:26:33   
24569  435007a5-6261-4d8b-b0a4-55fdc189754b         adore 2020-09-18 10:50:50   
24570  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-10-31 03:58:44   
24571  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-06-25 15:12:29   
24572  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-12-17 16:32:57   

      Content_Type        Category Sentiment  Score  
0            photo        studying  negative      0  
1            photo        studying  negative      0  
2            photo        studying  negative      0  
3            photo  healthy eating  negative      0  
4            photo  healthy eating  negative      0  
...            ...             ...       ...    ...  
24568        audio        veganism  positive     72  
24569        audio        veganism  positive     72  
24570          GIF         culture  positive     72  
24571          GIF         culture  positive     72  
24572          GIF         culture  positive     72  

[24573 rows x 7 columns]

* We have over 24,573 rows of data. Let's use this table to answer our business question regarding the top five categories with the largest aggregate popularity.

# Top Category Analysis

* The measure of popularity provided to us by the client is the "Score" column. To find the highest aggregate popularity for each category, we simply have to sum this value by category.

In [ ]:
#group by category and calculate total score
top_five = top_performers.groupby('Category').sum().sort_values(by='Score',ascending=False)
top_five.head(5)

<ipython-input-78-1813259fd6fe>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  top_five = top_performers.groupby('Category').sum().sort_values(by='Score',ascending=False)


Score
Category             
animals         74965
science         71168
healthy eating  69339
technology      68738
food            66676

* There's our answer - the top five categories by aggregate popularity are animals, science, healthy eating, technology, and food. We can export this cleaned dataset to a CSV to share with the client.



In [ ]:
top_performers.to_csv('client_data.csv')